### PR TITLE
dict_array and dict_map container classes added.

### PR DESCRIPTION
**Pull request by azhogin**

### CODE REVIEW

File: stdlib/cpp-sdk/tvm/tuple_stack.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include <optional>
+#include <tvm/builder.hpp>
+#include <tvm/error_code.hpp>
+#include <tvm/assert.hpp>
+
+namespace tvm {
+
+template<class T>
```
Comment (akiramenai) [inline stdlib/cpp-sdk/tvm/tuple_stack.hpp]: Could you explain the purpose of this template class?

File: stdlib/cpp-sdk/tvm/dictionary.hpp [hunk 2]
```diff
@@ -23,40 +32,340 @@ class dictionary {
...
+
+  dictionary& dict_;
+  unsigned idx_;
+  std::optional<slice> sl_;
+
+  __always_inline Pair operator*() const {
+    require(!!sl_, error_code::iterator_overflow);
+    return {idx_, schema::parse<Element>(*sl_)};
+  }
+  //__always_inline reference operator*() {
```
Comment (akiramenai) [inline stdlib/cpp-sdk/tvm/dictionary.hpp]: Shall we remove the commented lines?

File: stdlib/cpp-sdk/tvm/dictionary.hpp [hunk 3]
```diff
@@ -23,40 +32,340 @@ class dictionary {
...
+
+template<class Element, unsigned KeyLen>
+struct ElementRef {
+  __always_inline ElementRef& operator=(Element elem) {
+    auto [sl, succ] = dict_.dictusetget(schema::build(elem).make_slice(), idx_, KeyLen);
+    if (!succ)
+      ++size_;
+    return *this;
+  }
+  /*operator Element() const {
```
Comment (akiramenai) [inline stdlib/cpp-sdk/tvm/dictionary.hpp]: Shall we remove the commented method?

File: stdlib/cpp-sdk/tvm/dict_base.hpp [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  Element get_or_throw(unsigned key, unsigned exception) const {
+    auto opt_val = lookup(key);
+    require(!!opt_val, exception);
+    return *opt_val;
+  }
+
+  // Returns {Key, Element, Success} to match the solidity array.min() function
+  //  (not {Element, Key, Success} as dictumin returns)
+  std::tuple<unsigned, Element, bool> min() const {
```
Comment (akiramenai) [inline stdlib/cpp-sdk/tvm/dict_base.hpp]: Do you know by a chance why Solidity team reordered elements here?

**Comment by akiramenai** [comment]
And btw, why not `dict_set` and `dict_map`?

File: stdlib/cpp-sdk/tvm/tuple_stack.hpp (hunk 1 above)
Comment (azhogin) [reply stdlib/cpp-sdk/tvm/tuple_stack.hpp:11]: `tuple_stack` allows to work with growing stack of supported elements. Comments added.

File: stdlib/cpp-sdk/tvm/dictionary.hpp (hunk 2 above)
Comment (azhogin) [reply stdlib/cpp-sdk/tvm/dictionary.hpp]: Thanks, removed.

File: stdlib/cpp-sdk/tvm/dictionary.hpp (hunk 3 above)
Comment (azhogin) [reply stdlib/cpp-sdk/tvm/dictionary.hpp]: Removed.

File: stdlib/cpp-sdk/tvm/dict_base.hpp (hunk 4 above)
Comment (azhogin) [reply stdlib/cpp-sdk/tvm/dict_base.hpp:41]: No.

**Pull request merged**